From f12e47cff94bb62f4e51d7cc8aefe8874c763f2e Mon Sep 17 00:00:00 2001
From: kendhia <kendhia@gmail.com>
Date: Mon, 24 Dec 2012 16:14:14 +0100
Subject: [PATCH] expand the ruby pkgwat wrapper to retrieve list of packages
 and package details

---
 lib/pkgwat.rb | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/lib/pkgwat.rb b/lib/pkgwat.rb
index f79a7fe..39e07fc 100644
--- a/lib/pkgwat.rb
+++ b/lib/pkgwat.rb
@@ -13,6 +13,7 @@ module Pkgwat
   DEFAULT_DISTROS = [F17, F16, EPEL6]
   PACKAGE_NAME = "rubygem-:gem"
   PACKAGES_URL = "https://apps.fedoraproject.org/packages/fcomm_connector/bodhi/query/query_active_releases"
+  PACKAGES_URL_LIST = "https://apps.fedoraproject.org/packages/fcomm_connector/xapian/query/search_packages"
 
   def self.check_gem(name, version, distros = DEFAULT_DISTROS, throw_ex = false)
     puts "Checking #{name} #{version}...\n"
@@ -41,6 +42,27 @@ module Pkgwat
     PACKAGE_NAME.gsub(":gem", gem)
   end
 
+  def self.total_rows(pattern)
+    JSON.parse(get_packages_paginated(pattern, 10, 0).body)["total_rows"]
+  end
+  
+  #this function queries and returns the specified number of packages starting at the specified row
+  def self.get_packages(pattern, start=0, num=nil) 
+    if num == nil
+      num = total_rows(pattern)
+    end
+    query = {"filters"=>{"search"=>pattern}, "rows_per_page"=>num, "start_row"=>start}
+    url = PACKAGES_URL_LIST + "/" + query.to_json
+    uri = URI.parse(URI.escape(url)) 
+    response = submit_request(uri)
+    parse_results(response.body)    
+  end 
+ 
+  #this function just queries for and returns a single package  
+  def self.get_package(pattern)
+    get_packages(pattern, 0, 1).first
+  end 
+  
   def self.search_params(gem)
     filters = { :package => package_name(gem) }
     { :filters => filters }
-- 
1.7.11.7

